src/server/pfs/server/driver.go

```diff
@@ -1380,103 +1402,351 @@ func (d *driver) flushRepo(ctx context.Context, repo *pfs.Repo) ([]*pfs.RepoInfo
+				}
+			}
+			if parent != nil {
+				commitInfo := &pfs.CommitInfo{}
+				if err := commits.Update(parent.ID, commitInfo, func() error {
+					// Add existing live commits in commitInfo.ChildCommits to the
+					// live children above lowestCommitInfo, then put them all in
+					// 'parent'
+					for _, child := range commitInfo.ChildCommits {
+						if _, ok := deleted[child.ID]; ok {
+							continue
+						}
+						liveChildren[child.ID] = struct{}{}
+					}
+					commitInfo.ChildCommits = make([]*pfs.Commit, 0, len(liveChildren))
+					for child := range liveChildren {
+						commitInfo.ChildCommits = append(commitInfo.ChildCommits, client.NewCommit(parent.Repo.Name, child))
+					}
+					return nil
+				}); err != nil {
+					return fmt.Errorf("err rewriting children of ancestor commit %v: %v", lowestCommitInfo.Commit, err)
+				}
 			}
 		}
+		return nil
+	}); err != nil {
+		return fmt.Errorf("error rewriting commit graph: %v", err)
 	}
 
-	// Delete the commit itself and subtract the size of the commit
-	// from repo size.
-	_, err = col.NewSTM(ctx, d.etcdClient, func(stm col.STM) error {
-		repos := d.repos.ReadWrite(stm)
-		repoInfo := new(pfs.RepoInfo)
-		if err := repos.Get(commit.Repo.Name, repoInfo); err != nil {
+	if deleteScratch {
+		// Delete the scratch space for this commit
+		// TODO put scratch spaces in a collection and do this in the txn above
+		prefix := d.scratchCommitPrefix(ctx, userCommit)
+		if _, err := d.etcdClient.Delete(ctx, prefix, etcd.WithPrefix()); err != nil {
 			return err
 		}
-		repoInfo.SizeBytes -= commitInfo.SizeBytes
-		repos.Put(commit.Repo.Name, repoInfo)
+	}
 
-		commits := d.commits(commit.Repo.Name).ReadWrite(stm)
-		return commits.Delete(commit.ID)
+	// Traverse affected repos and rewrite all branches so that no branch points
+	// to a deleted commit
+	// TODO this should happen in the txn as well, but can't because we have to
+	// list branches and commits in 'modifiedRepos', and you can't list in a txn
+	// 7) Collect all affected branches
+	affectedBranchInfos := make([]*pfs.BranchInfo, 0, (len(affectedRepos)*3)/2)
+	for repo := range affectedRepos {
+		repoBranchInfos, err := d.listBranch(ctx, client.NewRepo(repo))
+		if err != nil {
+			return err
+		}
+		for _, branchInfo := range repoBranchInfos {
+			if branchInfo.Head != nil {
+				if _, ok := deleted[branchInfo.Head.ID]; ok {
+					affectedBranchInfos = append(affectedBranchInfos, branchInfo)
+				}
+			}
+		}
+	}
+
+	// 8) sort affectedBranchInfos by subvenance so that we move upstream
+	// branches before downstream branches. This ensures that if createBranch
+	// creates a new output commit (because a deleted input's parent hasn't
+	// been processed yet either), then the new output commit will have the
+	// right provenance
+	sort.Slice(affectedBranchInfos, func(i, j int) bool {
+		return len(affectedBranchInfos[i].Provenance) < len(affectedBranchInfos[j].Provenance)
 	})
```

Comment:
Do you want to sort by subvenance or provenance here? The comment says subvenance but the code is doing provenance.

Reply:
Also in going from upstream to downstream is there a problem wherein a branch is still pointing at a deleted commit and gets triggered by `propagateCommit` from `createBranch` and thus creates a commit with a parent that doesn't exist because it uses the head of the branch as the parent commit.

Reply:
Let me add a test for this. I'm actually not sure if there's a solution to this that doesn't involve making `createBranch` a batch call (i.e. update a set of branches, then create a new "top layer" of commits if there are unprocessed inputs)